src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,257 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include "sd-device.h"
+
+#include "alloc-util.h"
+#include "cgroup-util.h"
+#include "devnum-util.h"
+#include "device-util.h"
+#include "main-func.h"
+#include "path-util.h"
+#include "pretty-print.h"
+#include "verbs.h"
+
+static int help(void) {
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        r = terminal_urlify_man("systemd-iocost", "1", &link);
+        if (r < 0)
+                return log_oom();
+
+        printf("%s [OPTIONS...]\n\n"
+               "Set up iocost model and qos solutions for block devices\n"
+               "\nCommands:\n"
+               "  apply <path> [solution]    Apply the known solution for the device, if any, otherwise does nothing\n"
+               "  query <path>               Query the known solution for the device\n"
+               "\nOptions:\n"
+               "  -h --help                  Show this help\n"
+               "     --version               Show package version\n"
+               "\nSee the %s for details.\n",
+               program_invocation_short_name,
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int get_known_solutions(sd_device *device, char ***solutions) {
+        const char *value;
+        int r;
+
+        r = sd_device_get_property_value(device, "IOCOST_SOLUTIONS", &value);
+        if (r < 0) {
+                return 1;
+        }
+
+        *solutions = strv_split(value, " ");
+        if (!*solutions)
+                return log_oom();
+
+        return 0;
```

Comment:
so as per conding style: we want that functions either succeed and then initialize all return params, or fail and then leave all return params untouched. Failure typically means a negative return code, and success a return code >= 0.

This rule is broken here in various ways. i.e. the strv_split() return value should be assigned to a temporary variable first, before it is written to `*solutions`, so that we don#t touch it when strv_split() fails due to oom.

Reply:
Also, in systemd we have this rule that functions should either log about all errors or about none (in which case the caller can log). Please make up your mind if this function shall be of the logging kind or of the non-logging kind. Right now it logs in OOM but not if sd_device_get_property_value() fails